File: DataStructures/Trees/BFS・DFS/atcoder/A65/A65.php
<?php

// 入力の読み込み
fscanf(STDIN, "%d", $N);
$A = array_map('intval', explode(' ', trim(fgets(STDIN))));

// 木構造（上司→部下の隣接リスト）を作成
$tree = array_fill(0, $N + 1, []);
for ($i = 0; $i < count($A); $i++) {
    $parent = $A[$i];   // 上司
    $child = $i + 2;     // 社員番号（2〜N）
    $tree[$parent][] = $child;
}

// 部下数を格納する配列
$subordinates = array_fill(0, $N + 1, 0);

// 再帰DFSで部下数をカウント
function dfs($node) {
    global $tree, $subordinates;
    $count = 0;
    foreach ($tree[$node] as $child) {
        $count += 1 + dfs($child);
    }
    $subordinates[$node] = $count;
    return $count;
}

dfs(1); // 社長からスタート

// 結果出力（1〜N）
echo implode(' ', array_slice($subordinates, 1)) . PHP_EOL;

File: DataStructures/Trees/BFS・DFS/atcoder/A65/A65.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var (
	tree        [][]int
	subordinates []int
)

func dfs(node int) int {
	count := 0
	for _, child := range tree[node] {
		count += 1 + dfs(child)
	}
	subordinates[node] = count
	return count
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	// 入力読み取り
	scanner.Scan()
	N, _ := strconv.Atoi(scanner.Text())

	tree = make([][]int, N+1)        // 1-indexed
	subordinates = make([]int, N+1)  // 結果格納用

	for i := 2; i <= N; i++ {
		scanner.Scan()
		a, _ := strconv.Atoi(scanner.Text())
		tree[a] = append(tree[a], i)
	}

	dfs(1) // 社長からDFS開始

	// 結果出力
	for i := 1; i <= N; i++ {
		fmt.Print(subordinates[i])
		if i < N {
			fmt.Print(" ")
		}
	}
	fmt.Println()
}
